### PR TITLE
perf(messaging, ios): Improve time to delivery of background messages on iOS

### PR DESCRIPTION
**Pull request by brianathere**
Convert didReceiveRemoteNotification to wait until the RN bridge initialized based on the completion of the call to setBackgroundMessageHandler by the client

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [ ] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

Tested delivery using Firebase messaging, messages with data, in the foreground and background on an actual device. Confirmed foreground notification delivery is still immediate. Confirmed background notification delivery occurs when the application is running but hidden, as well as when the application has been force quit.

As this requires an actual device and no existing test automation existed for these scenarios all of the testing was manual.

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Review by mikehardy — CHANGES_REQUESTED** [review]
Very interesting! I love this idea, where I will summarize the idea as "solve the iOS 'when is the RN bridge up' detection issue by using the 'setBackgroundMessageHandler' call as a positive signal to uncork messages'

Did I get that correct, as the idea?

If so, it seems like a nice fix and I'd like to collaborate with you to get this merged and released.

I see these things as necessary to get it in shape though

- nothing in the `tests/` directory should be included in this PR, they are not relevant to the fix and clutter the PR - please remove all those items
- the e2e test you changed is not executed on ios at all, and this is an ios fix, that change should  be reverted
- this is currently not backwards compatible, and as the `setBackgroundMessageHandler` API is marked as "android" it's quite possible that people are not calling it on iOS *but still receiving messages*. If I understand the code change proposed here, iOS will now hang indefinitely on message delivery if `setBackgroundMessageHandler` is not called. The previous time-delay/uncork logic needs to remain as a fallback I think, leaving your idea of the positive signal given by setBackgroundMessageHandler as a very nice optimization but without potentially causing problems for existing users.

What do you think?

**Comment by brianathere** [comment]
> Very interesting! I love this idea, where I will summarize the idea as "solve the iOS 'when is the RN bridge up' detection issue by using the 'setBackgroundMessageHandler' call as a positive signal to uncork messages'
> 
> Did I get that correct, as the idea?
> 
> If so, it seems like a nice fix and I'd like to collaborate with you to get this merged and released.
> 
> I see these things as necessary to get it in shape though
> 
> * nothing in the `tests/` directory should be included in this PR, they are not relevant to the fix and clutter the PR - please remove all those items
> * the e2e test you changed is not executed on ios at all, and this is an ios fix, that change should  be reverted
> * this is currently not backwards compatible, and as the `setBackgroundMessageHandler` API is marked as "android" it's quite possible that people are not calling it on iOS _but still receiving messages_. If I understand the code change proposed here, iOS will now hang indefinitely on message delivery if `setBackgroundMessageHandler` is not called. The previous time-delay/uncork logic needs to remain as a fallback I think, leaving your idea of the positive signal given by setBackgroundMessageHandler as a very nice optimization but without potentially causing problems for existing users.
> 
> What do you think?

Excellent feedback. I accidentally sent this up for review, well before I got started, so let me start again here in an hour with a working solution, and then let's go from there. 

I wasn't aware setBackgroundMessageHandler was considered optional on iOS, and further saw it marked android which didn't make complete sense since it is needed on iOS, so yeah, I'll need to give how to mark this as backward breaking or not a think based on feedback.

**Comment by mikehardy** [comment]
Hmm - I was keying that comment *mostly* off the fact the API call was marked `android` and not taken into account before. Perhaps I am fearing for backwards compatibility too much and/or incorrectly.

> I wasn't aware setBackgroundMessageHandler was considered optional on iOS

On a re-think, perhaps for this to matter at all, you would need a handler set? Stated differently: the only messages we are delaying here would be sent through the background message handler?

If so, then my assertion it is optional and needs management here is incorrect. I regret I'm not familiar enough with this exact area of code to know off-hand but it should be easy to verify the assertion, and I may be wrong...

**Comment by brianathere** [comment]
> Hmm - I was keying that comment _mostly_ off the fact the API call was marked `android` and not taken into account before. Perhaps I am fearing for backwards compatibility too much and/or incorrectly.
> 
> > I wasn't aware setBackgroundMessageHandler was considered optional on iOS
> 
> On a re-think, perhaps for this to matter at all, you would need a handler set? Stated differently: the only messages we are delaying here would be sent through the background message handler?
> 
> If so, then my assertion it is optional and needs management here is incorrect. I regret I'm not familiar enough with this exact area of code to know off-hand but it should be easy to verify the assertion, and I may be wrong...



> Hmm - I was keying that comment _mostly_ off the fact the API call was marked `android` and not taken into account before. Perhaps I am fearing for backwards compatibility too much and/or incorrectly.
> 
> > I wasn't aware setBackgroundMessageHandler was considered optional on iOS
> 
> On a re-think, perhaps for this to matter at all, you would need a handler set? Stated differently: the only messages we are delaying here would be sent through the background message handler?
> 
> If so, then my assertion it is optional and needs management here is incorrect. I regret I'm not familiar enough with this exact area of code to know off-hand but it should be easy to verify the assertion, and I may be wrong...

My latest updates are what I was intending to share.

I believe nothing has changed in the case of foreground delivery with my change, it just gets sent to the app.

In the background delivery will be held on a dispatch queue until the client code calls setBackgroundMessageHandler and provides a callback. If the client code never calls, the delivery request will wait indefinitely.

Since setBackgroundMessageHandler is used on iOS and Android I removed the platform flag from that method.

I looked into writing automation for testing this change, however since it requires a real device to test I tested in my application instead. I'v only tested on an iPhone 12 at this point, but will go back to an iPhone 7 before merging.

As I've not written ObjC in 25 years, I'd very much appreciate eyes on that code.

File: packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m
Comment (mikehardy) [inline packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m:75]: I'm afraid our logging is all over the place in the module, partly my fault.
But I believe the correct style is to use this shared utility:

> app/ios/RNFBApp/RNFBSharedUtils.h:#define DLog(fmt, ...) NSLog((@"%s [Line %d] " fmt), __PRETTY_FUNCTION__, __LINE__, ##__VA_ARGS__);

File: packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m
Comment (mikehardy) [inline packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:97]: This one would use a similar log macro but it's `ELog` in this case

File: packages/messaging/lib/index.js
Comment (mikehardy) [inline packages/messaging/lib/index.js:291]:
```suggestion
   * Set a handler that will be called when a message is received while the app is in the background. 
   * Should be called before the app is registered in `AppRegistry`, for example in `index.js`. 
   * An app is considered to be in the background if no active window is displayed.
```

File: packages/messaging/lib/index.js
Comment (mikehardy) [inline packages/messaging/lib/index.js:292]: Lifted from here https://rnfirebase.io/reference/messaging#setBackgroundMessageHandler

```suggestion
   * @param handler called with an argument of type messaging.RemoteMessage that must be async and return a Promise
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
I have some comments on logging and documentation but I don't have the Obj-C skills to know if I'm missing something. This is something that will affect the Notifee library a lot if there is an issue though, perhaps I could get a look from @helenaford ? For context if you have time Helena - the underlying idea seems sound to me, and the implementation in general even, to me - at this point I think we're just trying to suss out if if the code will actually spin-then-uncork messages correctly, with a side question (on my part) on what exactly will happen if no background handler is ever registered? (e.g., will it use 100% cpu, or ?)

File: packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m
Comment (brianathere) [reply packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m:75]: Awesome. Do you know, will this end up in the system console log? I needed for debugging in development to correlate the log lines with the other iOS system logs to match up when notifications arrived to the phone with when the app responded. If ELog/DLog only write to the react native log, not the device logs, these statements are far less interesting and I will just remove them.

Comment (mikehardy) [reply packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m:75]: Yes - they are wrapping NSLog so should be going to the same place.
If they were RCTLog calls they'd go to the react-native console but these should be in Console.app or Xcode native device output

Comment (brianathere) [reply packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m:75]: I see it wraps NSLog, perfect.

**Review by mikehardy — DISMISSED** [review]
This looks good to me know - as mentioned I myself am hoping for another set of eyes on the objective-c but I don't have any hangups personally and if no one else can take a look I'm biased towards merge.

Please note that this generates a patch set from the patch CI action (https://github.com/invertase/react-native-firebase/actions/runs/1069327151) - it generates the patch set as the full difference between current release (12.3.0 as of this typing) and current dev branch + this change so it contains more than just this PR, but stated differently it contains "everything approved for release plus the PR" so it's still worth trying to make sure it works for you in production via patch-package

In the absence of qualified objective-c review that's how I'll test integrate it in my app for verification as well, would love a report if you try it

**Comment by brianathere** [comment]
> I have some comments on logging and documentation but I don't have the Obj-C skills to know if I'm missing something. This is something that will affect the Notifee library a lot if there is an issue though, perhaps I could get a look from @helenaford ? For context if you have time Helena - the underlying idea seems sound to me, and the implementation in general even, to me - at this point I think we're just trying to suss out if if the code will actually spin-then-uncork messages correctly, with a side question (on my part) on what exactly will happen if no background handler is ever registered? (e.g., will it use 100% cpu, or ?)

When didReceiveRemoteNotification is called and the app is in the background and the javascript client has not yet set a background message handler, a task is queued on the GCD thread pool, and the task waits for the client to set a handler. The task shouldn't use CPU, but it will consume a thread from the thread pool. This could lead to resource (address space, allowed threads?) exhaustion, is a large number (millions? - I don't know enough about modern iOS to know what the limits will be) of didReceiveRemoteNotification events are delivered. To mitigate this I added a change to the wait to give up after 25 seconds, aligned with the timeout for the beginBackgroundTaskWithExpirationHandler block. I tested a bit, seems to work well.

**Comment by brianathere** [comment]
> This looks good to me know - as mentioned I myself am hoping for another set of eyes on the objective-c but I don't have any hangups personally and if no one else can take a look I'm biased towards merge.
> 
> Please note that this generates a patch set from the patch CI action (https://github.com/invertase/react-native-firebase/actions/runs/1069327151) - it generates the patch set as the full difference between current release (12.3.0 as of this typing) and current dev branch + this change so it contains more than just this PR, but stated differently it contains "everything approved for release plus the PR" so it's still worth trying to make sure it works for you in production via patch-package
> 
> In the absence of qualified objective-c review that's how I'll test integrate it in my app for verification as well, would love a report if you try it

Happy to try this, but I don't understand how to apply this patch. Can you point me to the documentation?

**Comment by mikehardy** [comment]
@brianathere oh sure! The main to know is that https://github.com/ds300/patch-package exists (and it's pretty easy to install), then having learned about patch-package, the patches.zip we generate makes sense - it has a few patches you drop in your `./patches/` directory and whenever you `yarn` (or `npm i` or whatever) they apply and that's that. A great way to carry local changes without forking a repo, or in our case here since we are a monorepo about the only way for people to easily integrate PRs here into their actual project

**Comment by mikehardy** [comment]
Unrelated: I just noticed the CLA wasn't signed yet either. Once things are reviewed I merge and release pretty quickly, so follow the details link on that pending check here and it'll help step through it

**Comment by brianathere** [comment]
> Unrelated: I just noticed the CLA wasn't signed yet either. Once things are reviewed I merge and release pretty quickly, so follow the details link on that pending check here and it'll help step through it

I've attempted signing the CLA multiple times, it keeps timing out. Will try again. Definitely would like to get this into our users hands this week.

**Comment by brianathere** [comment]
> > Unrelated: I just noticed the CLA wasn't signed yet either. Once things are reviewed I merge and release pretty quickly, so follow the details link on that pending check here and it'll help step through it
> 
> I've attempted signing the CLA multiple times, it keeps timing out. Will try again. Definitely would like to get this into our users hands this week.

This is what I get when I attempt to sign. Looks like I already have, but it still shows up as un-signed in the CI.

image

**Comment by brianathere** [comment]
> @brianathere oh sure! The main to know is that https://github.com/ds300/patch-package exists (and it's pretty easy to install), then having learned about patch-package, the patches.zip we generate makes sense - it has a few patches you drop in your `./patches/` directory and whenever you `yarn` (or `npm i` or whatever) they apply and that's that. A great way to carry local changes without forking a repo, or in our case here since we are a monorepo about the only way for people to easily integrate PRs here into their actual project

I've tested on an iPhone 7 on iOS 12.1, and on a 12 Max running 14.7, works well in my application in both locations. I see there is still a block on the CLA being signed, do I need to do something more to get this cleared and this change merged?

**Comment by mikehardy** [comment]
Sorry - just sweeping through notifications now (I do one big react-native-firebase pass a day, this is it). I do not believe there are any further actions required on your part.
I've asked for another set of Objective-C eyes on this just in case but I'm biased towards merge+release and will do so without review if I don't get a second look by this afternoon my time (that's a few hours from now).

**Review by mikehardy — APPROVED** [review]
Okay, just re-re-read this after the latest changes and it still LGTM. On a countdown for merge on my side, though it is possible we get a second review and they notice something I haven't

:+1:

**Comment by mikehardy** [comment]
I suppose the doc you just made better for #5545 is now no longer up to date, since you rooted out the problem entirely!

**Pull request merged**
